### PR TITLE
replace fnv with sha256, replace map with doubleCache, keeps two cach…

### PR DESCRIPTION
**Pull request by y0sher**
…es and retain the older one

### CODE REVIEW

**Review by barakshani — APPROVED** [review]

File: p2p/gossip/protocol.go [hunk 1]
```diff
@@ -18,14 +18,51 @@ const propagateHandleBufferSize = 1000 // number of MessageValidation that we al
 const ProtocolName = "/p2p/1.0/gossip"
 const protocolVer = "0"
 
-type hash uint32
+type hash [12]byte
```
Comment (antonlerner) [inline p2p/gossip/protocol.go:21]: I think we already have a definition for hash somewhere

File: p2p/gossip/protocol.go [hunk 2]
```diff
@@ -53,28 +86,24 @@ type Protocol struct {
...
 		Log:             log2,
 		config:          config,
 		net:             base,
 		localNodePubkey: localNodePubkey,
 		peers:           make(map[string]*peer),
 		shutdown:        make(chan struct{}),
-		oldMessageQ:     make(map[hash]struct{}), // todo : remember to drain this
-		relayQ:          relayChan,
-		messageQ:        make(chan protocolMessage, messageQBufferSize),
+		oldMessageQ:     newDoubleCache(10000), // todo : remember to drain this
```
Comment (antonlerner) [inline p2p/gossip/protocol.go]: use parameter or const for 10000

File: p2p/gossip/protocol.go [hunk 3]
```diff
@@ -18,14 +18,51 @@ const propagateHandleBufferSize = 1000 // number of MessageValidation that we al
...
+	size   uint
+	cacheA map[hash]struct{}
+	cacheB map[hash]struct{}
+}
+
+func newDoubleCache(size uint) *doubleCache {
+	return &doubleCache{size, make(map[hash]struct{}, size), make(map[hash]struct{}, size)}
+}
+
+func (a *doubleCache) lookupOrCreate(key hash) bool {
```
Comment (antonlerner) [inline p2p/gossip/protocol.go]: perhaps this should be called just lookup, the fact that it stores the value shouldn't matter... what do you think?

**Review by antonlerner — APPROVED** [review]

File: p2p/gossip/protocol.go (hunk 3 above)
Comment (y0sher) [reply p2p/gossip/protocol.go]: i actually think if this method was only lookup someone could mistakenly try to insert after looking up. though maybe a better name is needed I'll try to come up with it

File: p2p/gossip/protocol.go (hunk 1 above)
Comment (y0sher) [reply p2p/gossip/protocol.go:21]: the hash we have it [32]byte hash and this hash is only 12 bytes. it is better that we store it in a [12]byte type. also this hash is only for internal use of this package so no worries of misuse.

**Pull request merged**